### PR TITLE
Optimize standby event handling

### PR DESCRIPTION
**Pull request by zwim**
Here I want to optimize the standby behavior, by minimizing the usage of the `onEnterStandby` and `onLeaveStandby` handlers.

The goal is to eliminate most of that handlers and optimize (speed) up the rest.
This can be achieved by just shifting the execution time of scheduled tasks by ` - Device.last_standby_time` (notice the sign, which means they should be executed earlier.) 
So it is not necessary to unschedule related tasks in `onEnterStandby` and reschedule them in `onLeaveStandby`.
The savings in the unschedule part might be smaller than the savings in the rescheduling part. (We will save a table delete and a binary insert in every case.)

As I won't eliminate the `onXXXStandby` events, everything will work as usual. The only difference is, that scheduled tasks will get executed `last_standby_time` earlier, as the clock has _not_ ticked for exactly that `last_standby_time` _(this is what we want, anyway)._

This initial commit is just a POC with changes in the `autosuspend`-plugin together with header and footer (and a private userpatch). Seems to work so far.

- [x] `autosuspend`-plugin
- [x] plus `UIManager`
- [x] `readerfooter`
- [x] `readercoptlistener`
- [x] `autowarmth`-plugin 
- [x] `autoturn`-plugin
- [x] `autodim`-plugin 

As always: Throw thoughts on my brain until it gets muddy or find logical errors. ;-)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10203)

### CODE REVIEW

**Comment by NiLuJe** [comment]
On a purely theoretical level, I'm midly puzzled, because UIManager works on CLOCK_MONOTONIC, which does *not* tick during suspend, so I would naively assume that rewinding shceduled task by the amount of time spent in standby would just wreak havoc?

**Comment by NiLuJe** [comment]
> On a purely theoretical level, I'm midly puzzled, because UIManager works on CLOCK_MONOTONIC, which does _not_ tick during suspend/standby, so I would naively assume that rewinding shceduled task by the amount of time spent in standby would just wreak havoc?

Oh, wait, that's the whole point ;p. (I was mixing clock sources in my head, never mind me ;p).

**Comment by NiLuJe** [comment]
It appears GitHub ate my first comment (or I'm so tired that I messed up ;p), so:

I won't be able to properly review/test this until next week ;).

File: plugins/autosuspend.koplugin/main.lua [hunk 1]
```diff
@@ -637,22 +637,22 @@ function AutoSuspend:AllowStandbyHandler()
         -- Since we go straight to input polling, and that our time spent in standby won't have affected the already computed
         -- input polling deadline (because MONOTONIC doesn't tick during standby/suspend),
         -- tweak said deadline to make sure poll will return immediately, so we get a chance to run through the task queue ASAP.
-        -- This ensures we get a LeaveStandby event in a timely fashion,
-        -- even when there isn't actually any user input happening (e.g., woken up by the rtc alarm).
         -- This shouldn't prevent us from actually consuming any pending input events first,
         -- because if we were woken up by user input, those events should already be in the evdev queue...
         UIManager:consumeInputEarlyAfterPM(true)
+        -- If we are leaving this method we are sure, that the input polling deadline is zero (consumeInputEarly).
+        -- Then the input waiting is ecexuted, so it is save to schedule a new task.
```
Comment (NiLuJe) [inline plugins/autosuspend.koplugin/main.lua:644]: nit: typo in `executed`

**Comment by NiLuJe** [comment]
I've had the PowerCover break suspend in fun and interesting ways until a reboot, yeah...

**Review by Frenzie — APPROVED** [review]
lgtm, but I haven't tested it yet

**Comment by NiLuJe** [comment]
I've slacked off a bit, and only juuust started reinstalling my linux env, so I still haven't tried this either. Hopefully this weekend ;).

**Comment by NiLuJe** [comment]
> When playing with my Sage sometimes I get strange screen refreshes during reading. The refreshes look like that the screen is split into 4 parts, which are refreshed clockwise starting from the bottom left. Very fast, hard to follow.

That's a kernel issue. It appeared with the first kernel updates that touched the collision handling code (can't remember the W version), and subsequent tweaks haven't really managed to entirely squash it.

It's specific to the Sage; the Elipsa has a different collision handling (also borken, but slightly less user-visibly so).

**Comment by NiLuJe** [comment]
FWIW, I hadn't traced it to PM at all, as the main spot I was seeing it was actually inside Nickel; I've only very rarely seen in KOReader.

Then again, if I think about it, I think I added a slight delay to the screensaver display on sunxi because I was tripping another collision handling bug (I don't think it was this exact one, though, but I can't quite remember), so it makes sense.

**Comment by NiLuJe** [comment]
As for the Notification thing, that's interesting, but I'd be inclined to leave it alone: it's perfectly normal for PM to fail if it happens too soon after a screen update. We handle that gracefully by waiting for the timeout again, and since it's likely to be in the one digit range of seconds, it feels like a non-issue to me ;).

**Comment by NiLuJe** [comment]
(Still have'nt had time to test this; I'm hoping to deal with it tomorrow, but we'll probably merge this tomorrow regardless, as it looks perfectly sensible to me ;)).

**Comment by NiLuJe** [comment]
> As for the Notification thing

The NO_MERGE flag might help (because it bypasses said broken collision handling codepath), but we're probably already using it for Notifications because of my last attempt at dealing with the Libra 2 crashes?

(Can't check right now, though).

**Comment by NiLuJe** [comment]
Fair warning: I'll probably veto something like that, FWIW ;).

**Review by NiLuJe — CHANGES_REQUESTED** [review]
A few wording issues with some comments, but code looks good ;).

Speaking of comments, some stuff around `self.going_to_suspend` usage may also need to be reviewed (at least one of those instances mentions onLeaveStandby, which doesn't exist anymore).

> Reviewed 2 of 4 files at r1, 4 of 5 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10203#-:-NS7v9we3H4dDAiKzODm:b-6b9bqu)* status: all files reviewed, 4 unresolved discussions (waiting on @zwim)

___
*[`frontend/ui/uimanager.lua` line 249 at r3](https://reviewable.io/reviews/koreader/koreader/10203#-NS7tUbi-uIheC12yxVj:-NS7tUbi-uIheC12yxVk:bszfgn6) ([raw file](https://github.com/koreader/koreader/blob/88d77115eb901c21b4d2b9bd45f408efde49976b/frontend/ui/uimanager.lua#L249)):*
> ```Lua
> -- UIManager uses CLOCK_MONOTONIC (which doesn't tick during standby), so shifting the execution
> -- time by a negative value will lead to an execution at the expected time.
> -- @param time if positive execute the tasks later, if negative the should be executed earlier
> ```

nit: still a minor wording snafu left near the end there ;).
___
*[`plugins/autosuspend.koplugin/main.lua` line 644 at r3](https://reviewable.io/reviews/koreader/koreader/10203#-NS7uiYzAUuAVFx3NeZY:-NS7uiYzAUuAVFx3NeZZ:b-w87lky) ([raw file](https://github.com/koreader/koreader/blob/88d77115eb901c21b4d2b9bd45f408efde49976b/plugins/autosuspend.koplugin/main.lua#L644)):*
> ```Lua
>         UIManager:consumeInputEarlyAfterPM(true)
>         -- If we are leaving this method we are sure, that the input polling deadline is zero (consumeInputEarly).
>         -- Then the input waiting is executed, so it is save to schedule a new task.
> ```

Second sentence is a tad clunky, making it hard to grok. (So's the comma placement in the first one, but that's less problematic).
___
*[`plugins/autosuspend.koplugin/main.lua` line 650 at r3](https://reviewable.io/reviews/koreader/koreader/10203#-NS7uFfc4AwDmlUP88IY:-NS7uFfc4AwDmlUP88IZ:b-2oqbin) ([raw file](https://github.com/koreader/koreader/blob/88d77115eb901c21b4d2b9bd45f408efde49976b/plugins/autosuspend.koplugin/main.lua#L650)):*
> ```Lua
>         end
>     else
>         -- If we are leaving this method we are sure, that the input polling deadline fits the next scheduled task (deadline).
> ```

Comment slightly unclear. That's simply input deadline >= next task, right?

**Review by NiLuJe — CHANGES_REQUESTED** [review]
One final nit ;).

> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10203#-:-NSCHNO0BWK7DJQWsCyg:b1t2rhs)* status: all files reviewed, 1 unresolved discussion (waiting on @zwim)

___
*[`plugins/autosuspend.koplugin/main.lua` line 644 at r4](https://reviewable.io/reviews/koreader/koreader/10203#-NSCGlBo27c9Z2O9i6O6:-NSCGlBp5A8oUinBzFls:b-lpytqs) ([raw file](https://github.com/koreader/koreader/blob/5cad9e5812e14b4af50db5dcb952648ca6927911/plugins/autosuspend.koplugin/main.lua#L644)):*
> ```Lua
> 
>         -- When we exit this method, we are sure that the input polling deadline is zero (consumeInputEarly).
>         -- UIManagerger will check newly scheduled tasks before going to input polling again (with a new deadline).
> ```

nit: `UIManager` ;o).

**Comment by NiLuJe** [comment]
> Fair warning: I'll probably veto something like that, FWIW ;).

To expand a bit on that, I'm happy with the current behavior (i.e., that PM fails too soon after a screen update is perfectly normal, we should handle that properly (and we do)), so attempting to deal with this corner-case seems like gratuitous complexity at first glance.

On the other hand, if that happens to, as a byproduct, workaround a sunxi display driver issue, that's another kettle of fish, and I might be more lenient with this ;).

**Review by NiLuJe — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/10203)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @zwim)
